crates/factor-outbound-networking/src/runtime_config/spin.rs

```diff
@@ -0,0 +1,273 @@
+            ca_use_webpki_roots
+        } else {
+            // Use webpki roots by default *unless* explicit roots were given
+            ca_roots_file.is_none()
+        };
+
+        let root_certificates = ca_roots_file
+            .map(|path| self.load_certs(path))
+            .transpose()?
+            .unwrap_or_default();
+
+        let client_cert = match (client_cert_file, client_private_key_file) {
+            (Some(cert_path), Some(key_path)) => Some(super::ClientCertConfig {
+                cert_chain: self.load_certs(cert_path)?,
+                key_der: self.load_key(key_path)?,
+            }),
+            (None, None) => None,
+            (Some(_), None) => bail!("client_cert_file specified without client_private_key_file"),
+            (None, Some(_)) => bail!("client_private_key_file specified without client_cert_file"),
+        };
+
+        Ok(TlsConfig {
+            components,
+            hosts,
+            root_certificates,
+            use_webpki_roots,
+            client_cert,
+        })
+    }
+
+    // Parse certs from the provided file
+    fn load_certs(
+        &self,
+        path: impl AsRef<Path>,
+    ) -> io::Result<Vec<rustls_pki_types::CertificateDer<'static>>> {
+        let path = self.runtime_config_dir.join(path);
+        rustls_pemfile::certs(&mut io::BufReader::new(fs::File::open(path).map_err(
+            |err| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("failed to read cert file {:?}", err),
+                )
+            },
+        )?))
+        .collect::<io::Result<Vec<rustls_pki_types::CertificateDer<'static>>>>()
+    }
+
+    // Parse a private key from the provided file
+    fn load_key(
+        &self,
+        path: impl AsRef<Path>,
+    ) -> anyhow::Result<rustls_pki_types::PrivateKeyDer<'static>> {
+        let path = self.runtime_config_dir.join(path);
+        rustls_pemfile::private_key(&mut io::BufReader::new(
+            fs::File::open(path).context("loading private key")?,
+        ))
+        .map_err(|_| anyhow::anyhow!("invalid input"))
+        .transpose()
+        .ok_or_else(|| {
+            io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "private key file contains no private keys",
+            )
+        })?
+    }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct RuntimeConfigToml {
+    component_ids: Vec<spin_serde::KebabId>,
+    #[serde(deserialize_with = "deserialize_hosts")]
+    hosts: Vec<String>,
+    ca_use_webpki_roots: Option<bool>,
+    ca_roots_file: Option<PathBuf>,
+    client_cert_file: Option<PathBuf>,
+    client_private_key_file: Option<PathBuf>,
+}
+
+fn deserialize_hosts<'de, D: Deserializer<'de>>(deserializer: D) -> Result<Vec<String>, D::Error> {
+    let hosts = Vec::<String>::deserialize(deserializer)?;
+    for host in &hosts {
+        validate_host(host).map_err(serde::de::Error::custom)?;
```

Comment:
If we're going to validate the host later, is there really a need for it here?

Reply:
Validating here gives nice TOML errors.